### DISCUSSION TITLE
How can I record audio in Expo as a WAV file?

### DISCUSSION THREAD

**Discussion by aramis-it** [expo/expo#24034] [Archived: Questions]
I did not find any solution.

**Comment by sabuhiteymurov** [comment]
Expo Audio docs indicates that you can specify the supported file extensions for each platform (refer to the [Android](https://developer.android.com/guide/topics/media/platform/supported-formats) and [iOS](https://developer.apple.com/documentation/audiotoolbox/audio_file_stream_services/1576497-audio_file_types) documentation). When using ```Recording.createAsync```, set AndroidOutputFormat and AndroidAudioEncoder to ```DEFAULT```, file extension to ```.wav``` and IOSOutputFormat to ```LINEARPCM``` in  options. Here is the minimal example: 
[Gist](https://gist.github.com/sabuhiteymurov/bf93edb3ffd69d8e37b71a3ae393f0c6)

**Reply by Nesquik** [reply]
> Expo Audio docs indicates that you can specify the supported file extensions for each platform (refer to the [Android](https://developer.android.com/guide/topics/media/platform/supported-formats) and [iOS](https://developer.apple.com/documentation/audiotoolbox/audio_file_stream_services/1576497-audio_file_types) documentation). When using `Recording.createAsync`, set AndroidOutputFormat and AndroidAudioEncoder to `DEFAULT`, file extension to `.wav` and IOSOutputFormat to `LINEARPCM` in options. Here is the minimal example: [Gist](https://gist.github.com/sabuhiteymurov/bf93edb3ffd69d8e37b71a3ae393f0c6)

Are you positive this is working? 
I've reproduced this with Expo 51 and VLC is telling me the file is an AMR. I've also opened the file itself with an HEX editor and I don't see WAV headers

**Reply by davidgaag** [reply]
Did you figure this out? I don't have an Android with me right now to test with, and I really need to record to .wav for a live audio processing task.

**Reply by aramis-it** [reply]
I did not found direct solution. I used converter.

**Reply by ClaytonRoberts** [reply]
> I did not found direct solution. I used converter.

Thanks for the following up so far.  Could you share what converter you went with?  
Is it a client-side converter?

**Reply by aramis-it** [reply]
I don't remember exactly. I think I did it in server side.

**Reply by yybawang** [reply]
Android doesn't have wav/pcm options，So if you are doing work for live speech-to-text，try AAC format，This format works well for STT(of course you should check the remote server api)，It's works for me, hope help somebody

Example configuration:

const RecordingOptions = {
  isMeteringEnabled: true,
  android: {
    extension: '.aac',
    outputFormat: AndroidOutputFormat.AAC_ADTS,
    audioEncoder: AndroidAudioEncoder.AAC,
    sampleRate: 16000,
    numberOfChannels: 1,
    bitRate: 16000,
  },
  ios: {
    extension: '.wav',
    outputFormat: IOSOutputFormat.LINEARPCM,
    audioQuality: IOSAudioQuality.MEDIUM,
    sampleRate: 16000,
    numberOfChannels: 1,
    bitRate: 16000,
    // linearPCMBitDepth: 16,
    // linearPCMIsBigEndian: false,
    // linearPCMIsFloat: false,
  },
  web: {
    mimeType: 'audio/webm',
    bitsPerSecond: 128000,
  },
};

**Reply by suvam3000** [reply]
@yybawang I am working with  azure/ microsoft cognitive speech service STT. I tried different way of the config, that you have provided but it's not working. Most likely outputFormat has to be wav for andriod.

**Reply by Sabaa-Abdennabi** [reply]
still no solution till now ? i am facing the same problem .

**Comment by alextaikh** [comment]
Anyone knows an updated solution? Facing the same problem of needing wav for Azure STT

**Comment by admfrenchdev** [comment]
Been two days that I try doing this with whatever lib exists besides expo-av and still no result. Tbh really disappointed by myself ^^ 😢
Any help with this will be really appreciated.

**Comment by alextaikh** [comment]
So the only solution is to use aac and then pass that to Backend, and Backend convert acc to wave and send to Azure?

**Comment by admfrenchdev** [comment]
Found `@siteed/expo-audio-stream` that allows to record a WAV file.
This is working with Azure speechToText:
```js
import {
  AudioRecording,
  RecordingConfig,
  useAudioRecorder,
} from "@siteed/expo-audio-stream";
import {
  AudioConfig,
  SpeechTranslationConfig,
  TranslationRecognizer,
  SpeechConfig,
} from "microsoft-cognitiveservices-speech-sdk";
// [...component code]
  const { startRecording, stopRecording, isRecording } = useAudioRecorder({
    debug: true,
  });

  const onRecordStart = async () => {
    try {
      const { granted } = await Audio.requestPermissionsAsync();
      if (granted) {
        const config: RecordingConfig = {
          sampleRate: 16000,
          channels: 1,
          encoding: "pcm_16bit",
          onAudioStream: async (e) => {
            // This callback seems mandatory, even if you only want to record the file at the end
            console.log("data received");
          },
        };
        const result = await startRecording(config);
        console.log("Recording started with config:", result);
      }
    } catch (err) {
      console.error("Failed to start recording", err);
    }
  };

  const onRecordStop = async () => {
      const audioRecord = await stopRecording();

      // Ugly filepath trick, might be improved
      const blob = await fetch(uri.replace("file:/", "file:///")).then((res) =>
        res.blob()
      );
      const file = new File([blob], "record.wav");
      const audioConfig = AudioConfig.fromWavFileInput(file);

      const speechApiConfig = SpeechTranslationConfig.fromSubscription(
        // Be aware this subcription key shoudl never be public. Hardcoded here for POC purpose.
        // Use the authByToken way instead
        subKey,
        "eastus2"
      );

      speechApiConfig.speechRecognitionLanguage = "en-US";
      speechApiConfig.addTargetLanguage("fr-FR");
      speechApiConfig.addTargetLanguage("de-DE");

      const transRecognizer = new TranslationRecognizer(
        speechApiConfig,
        audioConfig
      );

      transRecognizer.recognized = (reco, e) => {
        console.log("Recognized text: ", e.result.text);
      };

      transRecognizer.recognizeOnceAsync(
        (result) => {
          console.log({
            originalText: result.text,
            translations: result.translations,
          });
        },
        (e) => {
          console.log("ERROR", e);
        }
      );

  }
// [...component code]
```
Thing that bothers me right now is that I can't make this work with `continuous recognition` forcing me to wait the end of recording to get translations.
Been trying to use the `onAudioStream` cb to send data during recording, but I might still be missing something as I get no response from Azure. Still in progress.

**Reply by ClaytonRoberts** [reply]
Great answer! Thanks for sharing this library (@siteed/expo-audio-stream). 
It took some wrangling for me to make it fit in with my existing code and libraries, but it does indeed record wave and it works quite well after getting it set up. [I should note, i've only deployed for iOS so far, but I think Android will work as well. just read the lib docs.] For recording in wave on the frontend, i think this lib is the answer if you're using Expo (i'm on ~51).  I was able to find other non-expo solutions, but they would have required ejecting to a bare rn app, and i'd lose the benefits of expo.  This library and your message share it really saved the day.

**Reply by a-rostami** [reply]
@ClaytonRoberts I was able to make it work on android with the same configuraiton but seems like its not working on ios and i get  the following error:
[CoreFoundation] *** Terminating app due to uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'Input HW format and tap format not matching'
have you installed any specific library other than @siteed/expo-audio-stream or configured any specific value in app.json? you'd do me a big favor!

**Reply by a-rostami** [reply]
@ClaytonRoberts BTW i'm using "expo": "~51.0.28" and "@siteed/expo-audio-stream": "^1.7.0" and i ran the test on ios 18.2 can you please also share the versions with me?

**Reply by ClaytonRoberts** [reply]
That's great news that you had success on Android. Here are some details that I hope may help.  But if it's like my experience, you may just have to keep playing with configs and really making sure they're set properly and in the right place etc.

on iOS, I had to get past some crashes every time the recorder would start up. and the error i was getting seems roughly similar to what you listed, but mine was for sampleRate not matching.  
You mentioned "with the same configuration".  I might be misunderstanding that... but I think the sampleRate must be different between iOS and Android. 48000 and 16000 respectively.   ...The thing that helped me was dialing in the configs using the docs on this page: https://deeeed.github.io/expo-audio-stream/docs/api-reference/recording-config/ 

these 3 seem to be the most important:
sampleRate: 48000,
channels: 1,
encoding: 'pcm_16bit',


Here are my configs that work on iOS. 

Config for the recording wrapper function (similar to 'handleStart' in the docs): 
const config = {
                sampleRate: 48000,  // 48kHz for iOS //todo: make this dynamic with 16000 for Android
                channels: 1,
                encoding: 'pcm_16bit',
                interval: 500,
                enableProcessing: true,
                pointsPerSecond: 1000,
                algorithm: 'rms',
                features: { 
                    energy: true, 
                    rms: true 
                },
                onAudioStream: async (event) => {
                    // console.log('Audio data:', event); //debug
                },
                onAudioAnalysis: async (data) => {
                    // console.log('Processing:', data); //debug
                }
            };
            
            
       And i have some settings in my plugins section of my app.config.js (which i'm using instead of app.json). 

       return {
        ...config,
        expo: {
        ...more code
                plugins: [
                ["@siteed/expo-audio-stream", {
                    audioContextConfig: {
                        sampleRate: 48000,
                        channelCount: 1,
                        echoCancellation: false,
                        noiseSuppression: false,
                        autoGainControl: true
                    }
                }],
                "expo-router",
                './plugins/withMicrophonePermission',
                [
                    "expo-av"
                ]
            ]
            
            ...more
            
----
Versions: 
"expo": "~51.0.18", , "@siteed/expo-audio-stream": "^1.6.1",  ipad iOS 17.4.1 (deployed) and 17.5 (simulated on macbook) [not intentionally chosen iOS versions. Just happened to be those. ]

----
also, sounds like you're past this but i was hung up for a bit on running npx expo prebuild  after updating my configs. and then switching to npx expo run:ios for running the app rather than npx expo start which i had been using before adding this library.

**Reply by a-rostami** [reply]
@ClaytonRoberts thank you so much for the detailed answer! i got it working by setting sampleRate to 48000 on ios. my config is like this:
const config: RecordingConfig = {
                sampleRate: Device.osName == 'Android' ? 16000 : 48000,
                channels: 1,
                onAudioStream: async (e) => {
                  // This callback seems mandatory, even if you only want to record the file at the end
                  console.log("data received");
                }
            }
            
and yes seems like according to this issue https://github.com/expo/expo/issues/32982 expo-audio can not be used with expo go

**Reply by ClaytonRoberts** [reply]
Great to hear it!  Glad you got it working!

**Reply by Sabaa-Abdennabi** [reply]
can you please specify which version of audio stream and expo you're using because i am using expo 51.0.23 and audio stream 1.7.0 and i am facing this problem  Error: Cannot find native module 'ExpoAudioStream', js engine: hermes (tried deleting nodes , building ,clearing cache still doesn't work ) . I'd appreciate any kind of help !

**Reply by ClaytonRoberts** [reply]
@Sabaa-Abdennabi, "expo": "~51.0.18", "@siteed/expo-audio-stream": "^1.6.1", 

But i think the issue you're facing is related to how you're building. 
The docs mention using prebuild (https://deeeed.github.io/expo-audio-stream/docs/installation/). And i figured out after facing the same issue you are that when you're using react native rather than expo modules, you need to build in a different way. 
This expo-audio-stream library - i assume - is using RN libraries that necessitate a different build approach than what expo and hermes etc do.  So the way around the issue is to prebuild, pod install, and expo run:ios. It's a fuller build than what expo start does. 
I have Simulator set up to run ipad/iphone emulation. and run:ios leverages that.

**Reply by ClaytonRoberts** [reply]
here's an appstore note that will save you some hours: 
I think the expo-audio-stream affects the pod install and adds UIBackgroundModes permissions to the Info.plist under the ios/[appName] folder.  the app store reviewer i had noticed that my app was asking for audio permissions in background, but that the app didn't use them. and that was a holdup on approval.  ...To fix it, after the prebuild but before the export of the app to the app store, i manually removed the audio item from the plist file. ...I think i'll have to do this each time, bc my efforts to do it programmatically failed; Programatic might be possible, but i couldn't pull it off.

**Reply by Sabaa-Abdennabi** [reply]
Thank you so much @ClaytonRoberts this helped a lot . I was using expo go which caused that error to appear . Once i built the app using npx expo prebuild and npx expo run:android it worked just fine . THANKSS  !

**Reply by Chryogenic** [reply]
> Found `@siteed/expo-audio-stream` that allows to record a WAV file. This is working with Azure speechToText:
> 
> ```js
> import {
>   AudioRecording,
>   RecordingConfig,
>   useAudioRecorder,
> } from "@siteed/expo-audio-stream";
> import {
>   AudioConfig,
>   SpeechTranslationConfig,
>   TranslationRecognizer,
>   SpeechConfig,
> } from "microsoft-cognitiveservices-speech-sdk";
> // [...component code]
>   const { startRecording, stopRecording, isRecording } = useAudioRecorder({
>     debug: true,
>   });
> 
>   const onRecordStart = async () => {
>     try {
>       const { granted } = await Audio.requestPermissionsAsync();
>       if (granted) {
>         const config: RecordingConfig = {
>           sampleRate: 16000,
>           channels: 1,
>           encoding: "pcm_16bit",
>           onAudioStream: async (e) => {
>             // This callback seems mandatory, even if you only want to record the file at the end
>             console.log("data received");
>           },
>         };
>         const result = await startRecording(config);
>         console.log("Recording started with config:", result);
>       }
>     } catch (err) {
>       console.error("Failed to start recording", err);
>     }
>   };
> 
>   const onRecordStop = async () => {
>       const audioRecord = await stopRecording();
> 
>       // Ugly filepath trick, might be improved
>       const blob = await fetch(uri.replace("file:/", "file:///")).then((res) =>
>         res.blob()
>       );
>       const file = new File([blob], "record.wav");
>       const audioConfig = AudioConfig.fromWavFileInput(file);
> 
>       const speechApiConfig = SpeechTranslationConfig.fromSubscription(
>         // Be aware this subcription key shoudl never be public. Hardcoded here for POC purpose.
>         // Use the authByToken way instead
>         subKey,
>         "eastus2"
>       );
> 
>       speechApiConfig.speechRecognitionLanguage = "en-US";
>       speechApiConfig.addTargetLanguage("fr-FR");
>       speechApiConfig.addTargetLanguage("de-DE");
> 
>       const transRecognizer = new TranslationRecognizer(
>         speechApiConfig,
>         audioConfig
>       );
> 
>       transRecognizer.recognized = (reco, e) => {
>         console.log("Recognized text: ", e.result.text);
>       };
> 
>       transRecognizer.recognizeOnceAsync(
>         (result) => {
>           console.log({
>             originalText: result.text,
>             translations: result.translations,
>           });
>         },
>         (e) => {
>           console.log("ERROR", e);
>         }
>       );
> 
>   }
> // [...component code]
> ```
> 
> Thing that bothers me right now is that I can't make this work with `continuous recognition` forcing me to wait the end of recording to get translations. Been trying to use the `onAudioStream` cb to send data during recording, but I might still be missing something as I get no response from Azure. Still in progress.

Any luck on getting it to work with   `continuous recognition`?

**Comment by Sabaa-Abdennabi** [comment]
I personally made a format converter using python using ffmpeg to convert to .wav because i found no other alternative to record using .wav on android

**Reply by admfrenchdev** [reply]
Well, if it works, it works!
But I personnaly prefer using user device to compute instead of paying a server to do the job (without speaking about latency...) ;)

**Reply by Sabaa-Abdennabi** [reply]
that's true i have faced a latency issue that's why i am working on finding another alternative , thank you for the opinion !

**Comment by OliverHuang-WTG** [comment]
Found another way to solve this problem
I used `recorder-core` to convert amrnb format to pcm format (wav can also be converted)

This is the config of `useAudioRecorder` :
```
{
  extension: ".caf",
  numberOfChannels: 1,
  sampleRate: 16000,
  android: {
    extension: ".amr",
    outputFormat: "amrnb",
    audioEncoder: "amr_nb",
  },
  ios: {
    outputFormat: IOSOutputFormat.LINEARPCM,
    audioQuality: AudioQuality.MAX,
    linearPCMBitDepth: 16,
    linearPCMIsBigEndian: false,
    linearPCMIsFloat: false,
  },
};
```

Then I need to translate amrnb to pcm:
```
import * as FileSystem from "expo-file-system";
import "recorder-core";
import "recorder-core/src/engine/beta-amr";
import "recorder-core/src/engine/beta-amr-engine";
import "recorder-core/src/engine/pcm";
import Recorder from "recorder-core";
import { Platform } from "react-native";

function int16ArrayToBase64(array: Int16Array): string {
  const bytes = new Uint8Array(array.buffer);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

function sampleRaise(
  pcmDatas: Int16Array[],
  pcmSampleRate: number,
  newSampleRate: number
): Int16Array {
  let size = 0;
  for (let i = 0; i < pcmDatas.length; i++) {
    size += pcmDatas[i].length;
  }

  let step = newSampleRate / pcmSampleRate;
  if (step <= 1) {
    step = 1;
    newSampleRate = pcmSampleRate;
  } else {
    size = Math.floor(size * step);
  }

  let filterFn: any = 0;
  if (pcmSampleRate <= (newSampleRate * 3) / 4) {
    filterFn = Recorder.IIRFilter(
      true,
      newSampleRate,
      ((pcmSampleRate / 2) * 3) / 4
    );
  }

  const res = new Int16Array(size);

  let posFloat = 0,
    prev = 0;
  let F = filterFn && filterFn.Embed,
    Fx = 0,
    Fy = 0;
  for (let index = 0, nl = pcmDatas.length; index < nl; index++) {
    let arr = pcmDatas[index];
    for (let i = 0; i < arr.length; i++) {
      const cur = arr[i];

      let pos = Math.floor(posFloat);
      posFloat += step;
      const end = Math.floor(posFloat);

      const n = (cur - prev) / (end - pos);
      for (let j = 1; pos < end; pos++, j++) {
        //res[pos]=cur;
        let s = Math.floor(prev + j * n);
        if (F) {
          Fx = s;
          Fy =
            F.b0 * Fx + F.b1 * F.x1 + F.b0 * F.x2 - F.a1 * F.y1 - F.a2 * F.y2;
          F.x2 = F.x1;
          F.x1 = Fx;
          F.y2 = F.y1;
          F.y1 = Fy;
          s = Fy;
        } else {
          s = filterFn ? filterFn(s) : s;
        }

        if (s > 0x7fff) s = 0x7fff;
        else if (s < -0x8000) s = -0x8000;
        res[pos] = s;
      }

      prev = cur;
    }
  }

  return res;
}

function amr2pcm(base64File: string): Promise<string> {
  const bstr = atob(base64File);
  let n = bstr.length;
  const u8arr = new Uint8Array(n);
  while (n--) {
    u8arr[n] = bstr.charCodeAt(n);
  }

  return new Promise(async (resolve, reject) => {
    Recorder.AMR.decode(
      u8arr,
      function (pcm: ArrayBuffer) {
        const rec = Recorder({
          type: "pcm",
          sampleRate: 8000,
          bitRate: 16,
        });
        rec.dataType = "arraybuffer";
        rec.mock(pcm, 8000).stop(async function (wavBlob: ArrayBuffer) {
          resolve(
            int16ArrayToBase64(
              sampleRaise([new Int16Array(wavBlob)], 8000, 16000)
            )
          );
        }, reject);
      },
      reject
    );
  });
}

export async function test(uri: string) { // this is a test method to call amr2pcm
  const file = await FileSystem.readAsStringAsync(`file://${uri}`, {
    encoding: FileSystem.EncodingType.Base64,
  });
  if (Platform.OS === "android") {
    const pcm = await amr2pcm(file);
    return pcm;
  }

  return file;
}

```

If you want to convert to wav, you can import `recorder-core/src/engine/wav`.
then set the type to `wav` when new `Recorder`.

Of course, you can also convert pcm to wav directly, because wav only has one more header than pcm.

**Comment by ishmamx** [comment]
Was anyone able to make it work with Google or Microsoft real time transcription (voice to text)? (currently my app does the transcription when recording stops, but I want real time) :(
